src/sync_backend_clients/gitlab_sync_backend_client.js

```diff
@@ -0,0 +1,314 @@
+      .map((it) => ({
+        id: it.id,
+        name: it.name,
+        // Organice requires a leading "/", whereas GitLab API doesn't use one.
+        path: `/${it.path}`,
+        isDirectory: isDirectory(it),
+      }))
+      .sort((a, b) => {
+        // Folders first.
+        if (a.isDirectory && !b.isDirectory) {
+          return -1;
+        } else if (!a.isDirectory && b.isDirectory) {
+          return 1;
+        } else {
+          // Can't have same name, so don't need to check if equal/return 0.
+          return a.name > b.name ? 1 : -1;
+        }
+      })
+  );
+};
+
+const API_URL = 'https://gitlab.com/api/v4';
+
+/**
+ * GitLab sync backend, implemented using their REST API.
+ *
+ * @see https://docs.gitlab.com/ee/api/api_resources.html
+ * @param {OAuth2AuthCodePKCE} oauthClient
+ */
+export default (oauthClient) => {
+  const decoratedFetch = oauthClient.decorateFetchHTTPClient(fetch);
+
+  const getProjectApi = () => `${API_URL}/projects/${getPersistedField('gitLabProject')}`;
+
+  const isSignedIn = async () => {
+    if (!oauthClient.isAuthorized()) {
+      return false;
+    }
+    // Verify that we have an OAuth token (and refresh if needed). Don't care about return value,
+    // because the library handles persisting for us.
+    try {
+      await oauthClient.getAccessToken();
+      return true;
+    } catch {
+      return false;
+    }
+  };
+
+  /**
+   * Check that project exists and user is *probably* able to commit to it. This doesn't take branch
+   * protection into consideration, so it's not perfect... but who uses protected branches for the
+   * org files?
+   *
+   * This is separate from `isSignedIn` to avoid the overhead of multiple API calls every time the
+   * page is loaded.
+   */
+  const isProjectAccessible = async () => {
+    // Check project exists and user is a member who can *probably* commit.
+    const [userResponse, membersResponse] = await Promise.all([
+      // https://docs.gitlab.com/ee/api/users.html#list-current-user-for-normal-users
+      decoratedFetch(`${API_URL}/user`),
+      // https://docs.gitlab.com/ee/api/members.html#list-all-members-of-a-group-or-project
+      decoratedFetch(`${getProjectApi()}/members`),
+    ]);
+    if (!userResponse.ok || !membersResponse.ok) {
+      return false;
+    }
+    const [user, members] = await Promise.all([userResponse.json(), membersResponse.json()]);
+    const matched = members.find((m) => m.id === user.id);
+    // Access levels: https://docs.gitlab.com/ee/api/members.html#valid-access-levels
+    // Permissions: https://docs.gitlab.com/ee/user/permissions.html#project-members-permissions
+    // 30 is developer, which is the minimum to commit to a non-protected branch. If branch is
+    // protected then they still might be be able to commit, but this is good enough.
+    return matched && matched.access_level >= 30;
+  };
+
+  let cachedDefaultBranch;
+  const getDefaultBranch = async () => {
```

Comment:
Good point. What if the user changes their default branch and it's still cached? Then, the subsequent API actions would yield errors like "not found" and the user have to logout/login, right?

Reply:
My 2c: Lots of tools fail when the repo admin changes the default branch. When some people switched to `main`, we've seen errors in npm packages, Clojure modules, CI and so forth and so on.

Personally, I'd cache the branch instead of making a conditional API request ahead of every other API request. If the user changes the default branch they can re-login. Or, we could add a dropdown in the settings where she can select the branch she wants to work on. Since the first option is easier, I'd personally go with that one and leave the latter to the user who wants to switch branches.

Reply:
> What if the user changes their default branch and it's still cached? Then, the subsequent API actions would yield errors like "not found" and the user have to logout/login, right?

Not quite. The `cachedDefaultBranch` is just a local variable scoped to the client object. It isn't persisted to `localStorage` or anything, so all the user has to do is refresh the page. Since this is a SPA that means it is only fetched once so long as they keep the browser tab open, don't refresh the page, and the browser doesn't unload the page due to low memory.

> Personally, I'd cache the branch instead of making a conditional API request ahead of every other API request. If the user changes the default branch they can re-login.

@munen I am caching it, but only in a local variable instead of persisting to storage -- see previous paragraph. It seems like a good compromise of avoiding too many API calls without causing edge cases from persisting it long-term like what @schoettl was getting at. I would prefer to avoid persisting to `localStorage` until we add a dropdown in settings like you mention, out of concern for edge cases that are hard to convey to the user.

If that explanation doesn't make sense, you might check out the network requests tab of your browser's dev tools to see what I mean.

So does leaving this the way it is currently sound good, or are you suggesting a different form of caching?

Reply:
> So does leaving this the way it is currently sound good, or are you suggesting a different form of caching?

Works for me! I didn't say so explicitly, but when I said "I'd cache the branch" I already did want to agree with your existing strategy which I saw from the code. I could have been clearer in that. 

In any case, kudos to the original implementation and thank you for the additional explanation for both of us :pray:

Reply:
@schoettl Will you resolve this conversation if you're fine with the current implementation, please?